### ISSUE TITLE
[Question] Moving to psr/http-client

### ISSUE THREAD

**Issue by Gounlaf** [mcustiel/phiremock-client#1]
Hi @mcustiel,

Would you agree to move to `psr/http-client` instead of guzzle specifically?
It would give more flexibility and less conflict when using phiremock.

I think I can help you doing the move if you need.

Regards,

**Comment by mcustiel** [comment]
Hello @Gounlaf 
For me it would be OK to be more general than specific. Could you explain me what do you have in mind? Right now I need Guzzle because it's the default implementation, and the client class does not expect guzzle specifically but a RemoteConnectionInterface object. Is your idea to provide a default implementation without guzzle?

**Comment by Gounlaf** [comment]
> Right now I need Guzzle because it's the default implementation

So `\Mcustiel\Phiremock\Client\Factory` is the "default factory" ?
If so, guzzle should be required as a `suggested dependency` + `required-dev dependency` instead of a required dependency, avoiding installing by default Guzzle.

Anyway, looking at `\Mcustiel\Phiremock\Common\Http\Implementation\GuzzleConnection`, it could be replaced with a "PsrConnection" (accepting a psr/http-client class in ctor), making it fully agnostic.

**Comment by mcustiel** [comment]
I can implement a pure CURL client, because a default implementation is needed. My concern is that by not providing a default implementation in the client, I am making it more complicated to set up the client and the codeception module. I can then extend the module to provide an implementation of the http client and pass it to the client.

**Comment by Gounlaf** [comment]
There are some "discovery" class that can find & configure default client automatically.
IMO you can/should provide a default connection as you do now. The main issue - for me as a user - is that Guzzle **is** required. That's why I said:

> If so, guzzle should be required as a suggested dependency + required-dev dependency instead of a required dependency, avoiding installing by default Guzzle.

Give me until this weekend to provide you a "PsrConnection" and "GuzzleConnection" compatible with it (Guzzle 7.x is compatible Psr18) ;)

**Comment by Gounlaf** [comment]
It should work: https://github.com/mcustiel/phiremock-common/pull/3

**Comment by mcustiel** [comment]
Awesome. I had a short look at the PR and it looks good. I will take that and still provide a default pure-php implementation with curl. I want that everything just works after installation. I will do it this weekend. Thank you very much @Gounlaf

**Comment by Gounlaf** [comment]
> I will take that and still provide a default pure-php implementation with curl. I want that everything just works after installation.

I think you shouldn't do it.

In the README / Install, just put something like this:
**TL;DR:** `composer require mcustiel/phiremock guzzlehttp/guzzle`

Does it worth the time you will spend on making a curl implementation?
Before, you relied on guzzle; you can still rely on it as a "default" implementation. You were requiring explicitly a dependency. Now, it's the user choice to require it.

And, by the way, `curl` is "an option". It's usually packaged - the php packages - separately, and curl is not installed by default on distros. If I'm right Guzzle (or other clients) will use curl if available, and fallback on `streams`.
So, by providing a curl implementation by default, you will now have to require `curl`. It's a bigger constraint.

Sorry if it looks mean or rude.

**Comment by mcustiel** [comment]
> Sorry if it looks mean or rude.

Not at all, actually I am glad to have someone to help me think about this. I am still getting my mind around this and I already came to the conclusion that curl is also a fixed dependency. 
Then I think you have a point, I will take your changes and continue working on top of them during the weekend.
Thank you very much.
